src/main/python/mmlspark/cyber/ml/access_anomalies/collaborative_filtering.py

```diff
@@ -0,0 +1,721 @@
+        self.output_col = output_col
+
+    @property
+    def tenant_col(self):
+        return self.user_res_cf_df_model.tenant_col
+
+    @property
+    def user_col(self):
+        return self.user_res_cf_df_model.user_col
+
+    @property
+    def user_vec_col(self):
+        return self.user_res_cf_df_model.user_vec_col
+
+    @property
+    def res_col(self):
+        return self.user_res_cf_df_model.res_col
+
+    @property
+    def res_vec_col(self):
+        return self.user_res_cf_df_model.res_vec_col
+
+    @property
+    def user_model_df(self):
+        return self.user_res_cf_df_model.user_model_df
+
+    @property
+    def res_model_df(self):
+        return self.user_res_cf_df_model.res_model_df
+
+    def _transform(self, df: DataFrame) -> DataFrame:
+        dot = _make_dot()
+
+        tenant_col = self.tenant_col
+        user_col = self.user_col
+        user_vec_col = self.user_vec_col
+        res_col = self.res_col
+        res_vec_col = self.res_vec_col
+        output_col = self.output_col
+
+        return df.join(
+            self.user_model_df, [tenant_col, user_col], how='left'
+        ).join(
+            self.res_model_df, [tenant_col, res_col], how='left'
+        ).withColumn(
+            output_col,
+            dot(f.col(user_vec_col), f.col(res_vec_col))
+        ).drop(
+            user_vec_col
+        ).drop(
+            res_vec_col
+        )
+
+
+class CfAlgoParams:
+    """
+    Parameter for the AccessAnomaly Estimator indicating
+    if to use implicit or explicit feedback versions of the ALS algorithm
+    and additional parameters that are relevant for the implicit/explicit versions
+    """
+    def __init__(self, implicit: bool):
+        self._implicit = implicit
+        self._alpha: Optional[float] = None
+        self._complementset_factor: Optional[int] = None
+        self._neg_score: Optional[float] = None
+
+        # set default values
+        if implicit:
+            self.set_alpha(1.0)
+        else:
+            self.set_complementset_factor(2)
+            self.set_neg_score(1.0)
+
+    @property
+    def implicit(self) -> bool:
+        return self._implicit
+
+    @property
+    def alpha(self) -> Optional[float]:
+        """
+        Relevant for the implicit version only
+        :return: alpha value (see descriptions in explainParam of AccessAnomaly)
+        """
+        return self._alpha
+
+    @property
+    def complementset_factor(self) -> Optional[int]:
+        """
+        Relevant for the explicit version only
+        :return: complementset_factor value (see descriptions in explainParam of AccessAnomaly)
+        """
+        return self._complementset_factor
+
+    @property
+    def neg_score(self) -> Optional[float]:
+        """
+        Relevant for the explicit version only
+        :return: neg_score value (see descriptions in explainParam of AccessAnomaly)
+        """
+        return self._neg_score
+
+    def set_alpha(self, value: float):
```

Comment:
canyou use SparkML params here and elsewhere in the code?

Reply:
Will be fixed by fixing CfAlgoParams

Reply:
done